### PR TITLE
Run block header validation e2e

### PR DESCRIPTION
**Pull request by m-kus**
This PR introduces `BlockHeader` structure which is a `Header` extended with the Merkle tree root. This structure is sufficient to do header-only validation and new `BlockHeaderValidator` trait allows to do that and update the chain state accordingly.

There is an accompanying integration test that checks `ChainState169 + BlockHeader170 = ChainState170`. This test has been written manually, but we should instead generate such tests via script, it will be useful to create regression tests in the future.

Finally, the main function is updated to perform block header validation, there is also a script that runs the program e2e, loading parameters from a json file and encoding them into field elements.

- [x] issue #96
- [x] follows contribution [guide](https://github.com/keep-starknet-strange/raito/blob/main/CONTRIBUTING.md)
- [x] code change includes tests

### CODE REVIEW

**Review by maciejka — COMMENTED** [review]
In general I agree that having the ability to validate headers alone in the project is a valid goal. What I worry about is having to maintain extra infrastructure. They way it is written we need an extra struct, validate function and a script to run it, extra test data, and a script to build this data. Overall a second execution path in the project which we will need to maintain.

I would prefer to incorporate header validation in the main flow even at the cost of some extra data and ifs in the code. How making `Block` dual use struct that would contain header and either a list of transactions or transaction hash? Validation function would become slightly more complex but we would save on infrastructure duplication.

WDYT?

**Comment by m-kus** [comment]
This is indeed a second path, but there are not just downsides :)

We can definitely make `Block` dual use struct and reuse (some) of the validation, but I'd argue that the current separation is cleaner and likely (I haven't measured though) reduces the # of steps, while merging two cases together will introduce more code. Maybe it's better to instead refactor the top level validation flow to reduce the overlap between full block/block header implementations?

As for the scripts and data - here I cannot think of a way to fully merge two cases. The key idea is that for block header validation you can do very lightweight requests with constant size response (probably we can even query batches from some data lake) and those responses are self-sufficient, i.e. no need to extend the data with other information (like we do for the full consensus client). But implementation wise it will likely to be the same Python file just with different CLI option.

File: scripts/data/felt_writer.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
+# This module provides helpers for serializing Python objects into Cairo runner arguments (felts and arrays).
```
[the remark this answers is not in the document]
Comment (m-kus) [reply scripts/data/felt_writer.py:1]: Yes, they provide ABI encoding/decoding functionality but that's specific to Starknet contracts.  
In our case we are cheating a bit by "looking under the hood" to see how types we are interested in are stored on the stack.

**Comment by maciejka** [comment]
> This is indeed a second path, but there are not just downsides :)
> 
> We can definitely make `Block` dual use struct and reuse (some) of the validation, but I'd argue that the current separation is cleaner and likely (I haven't measured though) reduces the # of steps, while merging two cases together will introduce more code. Maybe it's better to instead refactor the top level validation flow to reduce the overlap between full block/block header implementations?

I can afford to sacrifice some steps (very little compared to the cost sha). 
Dual use block will require less code. 
In fact we will need even more modes of operation when we introduce utreexo, I mean in the end there will be at least:
* header verification
* block verification without utreexo
* full verification
* maybe even: block verification without scripts and utreexo

I would frame header vs full block verification in this context. 

> As for the scripts and data - here I cannot think of a way to fully merge two cases. The key idea is that for block header validation you can do very lightweight requests with constant size response (probably we can even query batches from some data lake) and those responses are self-sufficient, i.e. no need to extend the data with other information (like we do for the full consensus client). But implementation wise it will likely to be the same Python file just with different CLI option.

If header case data script would use getheader instead of getblock + getrawtransactions.

Let's have a call to continue this discussion.

File: src/types/block.cairo [hunk 2]
```diff
@@ -8,12 +8,15 @@ use crate::utils::numeric::u32_byte_reverse;
 use super::transaction::Transaction;
 
 /// Represents a block in the blockchain.
-#[derive(Drop, Copy)]
+#[derive(Drop, Copy, Debug, PartialEq)]
 pub struct Block {
     /// Block header.
     pub header: Header,
     /// Transactions.
     pub txs: Span<Transaction>,
```
Comment (maciejka) [inline src/types/block.cairo:16]: This should be either `txid_root` or `txs`, an enum seems to be necessary.

**Review by maciejka — CHANGES_REQUESTED** [review]

File: scripts/data/light_client_args_170.json [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        ]
+    },
+    "block": {
+        "header": {
+            "version": 1,
+            "time": 1231731025,
+            "bits": 486604799,
+            "nonce": 1889418792
+        },
+        "txs": [],
```
Comment (maciejka) [inline scripts/data/light_client_args_170.json:32]: Seems redundant after recent changes.

**Review by maciejka — COMMENTED** [review]
get_block script is not updated because we want to supersede it with data based tests?

**Comment by m-kus** [comment]
> get_block script is not updated because we want to supersede it with data based tests?

Yes (but even if we want to still have integration tests in Cairo, it needs to be extended to construct chain state), left for a follow up task

**Review by maciejka — APPROVED** [review]

**Pull request merged**